peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+
+
+Specification
+=============
+
+The :external+py3.13:data:`typing.ReadOnly` :external+typing:term:`type qualifier`
+becomes a valid annotation for :term:`attributes <attribute>` of classes and protocols.
+It is used to indicate that an attribute should not be reassigned or ``del``\ eted.
+
+The deletability rule should be extended to ``Final`` as well, as it is currently
+not specified.
+
+Akin to ``Final``, read-only attributes do not influence the mutability of
+the assigned object. Immutable ABCs and containers may be used in combination with
+``ReadOnly`` to prevent mutation of such values.
+
+Syntax
+------
+
+``ReadOnly`` can be used at class-level or within ``__init__`` to mark individual
+attributes read-only:
+
+.. code-block:: python
+
+    class Book:
+        id: ReadOnly[int]
+
+        def __init__(self, id: int, name: str) -> None:
+            self.id = id
+            self.name: ReadOnly = name
+
+The explicit type in ``ReadOnly[<type>]`` can be omitted if the declaration has
+an initializing value. A type checker should apply its usual type inference
+rules to determine the type of ``name``.
+
+If an attribute is already implied to be read-only, like in the frozen :ref:`classes`,
+explicit declarations should be permitted and seen as equivalent, except ``Final``
+additionally forbids overriding in subclasses:
+
+.. code-block:: python
+
+    @dataclass(frozen=True)
+    class Point:
+        x: ReadOnly[int]
+        y: Final[int]
+
+Initialization
+--------------
+
+Assignment to a ``ReadOnly`` attribute can only occur as a part of the declaration,
+or within ``__init__`` of the same class. There is no restriction to how many
+times the attribute can be assigned to.
+
+.. code-block:: python
+
+    from collections import abc
+    from typing import ReadOnly
+
+
+    class Band:
+        name: str
+        songs: ReadOnly[list[str]]
+
+        def __init__(self, name: str, songs: abc.Iterable[str] | None = None) -> None:
+            self.name = name
+            self.songs = []
+
+            if songs is not None:
+                # multiple assignments during initialization are fine
+                self.songs = list(songs)
+
+        def clear(self) -> None:
+            # error: assignment to read-only "songs" outside initialization
+            self.songs = []
+
+
+    band = Band(name="Bôa", songs=["Duvet"])
+    band.name = "Python"  # ok: "name" is not read-only
+    band.songs = []  # error: "songs" is read-only
+    band.songs.append("Twilight")  # ok: list is mutable
+
+
+    class SubBand(Band):
+        def __init__(self) -> None:
+            # error: cannot assign to a read-only attribute of base class
+            self.songs = []
+
+An initializing value at a class level can serve as a `flyweight <https://en.wikipedia.org/wiki/Flyweight_pattern>`_
+default for instances:
+
+.. code-block:: python
+
+    class Patient:
+        number: ReadOnly = 0
```

Comment:
I think we should require `ReadOnly` to be explicitly parameterized in this case. Otherwise, a type checker has to look at `__init__` to figure out the type.

Reply:
I was worried about cases where the immediate initialization might not give complete information about the type (e.g.,
```
class C:
    x: ReadOnly = 0
    def __init__(self):
        self.x = 4.2
```
but if it's specified that the type is always inferred only from the RHS of the assignment regardless, I suppose that works.

Reply:
So are we trying to say that A is allowed and B is not allowed?

A:
```python
class C:
    x: Readonly[int] = 2

    def __init__(self):
        if random() < 0.1:
            self.x = 3
```

B:
```python
class C:
    x: Readonly = 2

    def __init__(self):
        if random() < 0.1:
            self.x = 3
```

(Because in B, the type could be inferred as `Literal[2]` so `3` would be invalid.)

Whatever we're saying, it could be made more clear.

Reply:
The current draft would allow both A and B. My suggestion was that we not allow B.

Reply:
There are several different ways that the code example in B could be "not allowed". One is that bare `ReadOnly` is not allowed as an annotation anywhere, another is that bare `ReadOnly` is not allowed as a class attribute annotation, and a third is that it is allowed here, but the type checker infers the type `Literal[2]`, and so the assignment of `3` in `__init__` is a type error.

The current draft allows the annotation, and leaves the inference unspecified, so the annotation in (B) would be accepted, and whether the assignment in `__init__` is accepted would depend on how the type checker chooses to do inference.

I (weakly) favor side-stepping all of this by just making bare `ReadOnly` never allowed anywhere. I understand the appeal of not requiring annotations that can be inferred, I just think that once you're writing a `ReadOnly` annotation, it's not that hard to specify the type, too, and it makes the behavior much less type-checker-dependent.

I think it will be a bit more confusing if bare `ReadOnly` is allowed in some places but not in others.

Reply:
I made it required everywhere.